datalad_next/runners/run.py

```diff
@@ -0,0 +1,64 @@
+"""
+"""
+
+from __future__ import annotations
+
+from contextlib import contextmanager
+from pathlib import Path
+from queue import Queue
+from subprocess import (
+    DEVNULL,
+    TimeoutExpired,
+)
+from typing import (
+    IO,
+)
+
+from datalad.runner.nonasyncrunner import _ResultGenerator
+
+from . import (
+    Protocol,
+    ThreadedRunner,
+)
+
+
+@contextmanager
+def run(
+    cmd: list,
+    protocol_class: Protocol,
+    *,
+    cwd: Path | None = None,
+    input: int | IO | bytes | Queue[bytes | None] | None = None,
+    # only generator protocols make sense for timeout, and timeouts are
```

Comment:
Timeouts make sense in non-generator execution as well, because they would trigger a call to the timeout-method of the protocol. The protocol implementation could then, for example, raise an exception.

Reply:
They make sense. But I found what you describe to not work/be implemented. I also found no test that would verify such behavior. I assume you write this from a perspective of a future development plan only.